hicommonwealth/commonwealth
# Upgrade Canvas packages to 0.10

This PR upgrades the version of Canvas we are using to the latest version (0.10.0).

To recap, our objective with these refactors is to produce signed sessions and actions that "prove" that an authenticated user performed some action (like creating a thread or a comment). These signed payloads can then be verified by anyone (for now this is just the backend) such as other users, which then lets us store all of this data in a decentralised fashion, instead of just trusting the backend server. Right now we are generating and signing these payloads, but with an old version of the Canvas libraries (0.5). We want to upgrade to the latest version, which changes a lot of the interface types and functions.

Changes to the Canvas library:
- We have introduced `SessionSigner` classes, this contains the logic for signing, verifying and storing (in localStorage) sessions. Previously a lot of this code lived in the web wallet classes, `verifyAddress` and the sessions controller. The result here is that we should be able to add new chains by adding new `SessionSigner` classes. It should also be much easier to add new web wallet support.
- `SessionPayload` and `ActionPayload` have been replaced with the `Session`, `Action` and `Message` classes.

The Commonwealth code has been changed in a number of places to reflect the changes in the Canvas library:
- [x] Delete the code for verifying sessions from `verifyAddress` and use the imported `SessionSigner` code
- [x] Replace the `canvas_action` and `canvas_session` fields from the database and API with a `canvas_signed_data` field. This is a JSON blob containing all of the data we need to verify actions. This means that if the Canvas message formats change, we don't need to update a huge amount of Commonwealth app code.
- [x] Update API tests

## QA plan

For each wallet:

- [x] cosmos evm metamask (evmos)
- [x] keplr ethereum web wallet (evmos) - Needs fix
- [x] keplr web wallet (cosmos)
- [x] leap web wallet (cosmos)
- [x] metamask web wallet (ethereum)
- [x] coinbase web wallet (ethereum)
- [x] walletconnect web wallet (ethereum)
- [x] polkadot web wallet (polkadot)
- [x] phantom web wallet (solana)
- [x] terra station web wallet (terra)
- [x] terra station mobile wallet (terra) - Needs fix (lower priority)
- [ ] cosmos magic wallet - Needs fix (lower priority)
- [x] ethereum magic wallet - Needs fix

Do the following actions:

- Log in using the wallet
- Create a thread
- Create a comment on the thread
- React to the thread
- React to the comment

You should be able to see the effects of these actions without any error messages and they should persist if you reload the page.

Then simulate the session expiring by deleting the session from localStorage (this should have a key like `canvas/canvas:commonwealth/eip155:1:0x45490B82C56f27B67Bf91811B49579F3155C4892`. Attempt to perform any action that requires authorisation and you should see a prompt to generate a new Canvas session.
